File: syncloud/src/main/java/org/syncloud/android/discovery/DiscoveryManager.java
package org.syncloud.android.discovery;

import android.net.nsd.NsdManager;
import android.net.wifi.WifiManager;

import org.apache.log4j.Logger;
import org.syncloud.android.discovery.nsd.NsdDiscovery;

public class DiscoveryManager {

    private static Logger logger = Logger.getLogger(DiscoveryManager.class.getName());

    private MulticastLock lock;

    private Discovery discovery;
    private NsdManager manager;
    private Boolean canceled = false;

    public static int ACTIVATION_PORT = 81;

    public DiscoveryManager(WifiManager wifi, NsdManager manager) {
        lock = new MulticastLock(wifi);
        this.manager = manager;
    }

    public void run(int timeoutSeconds, DeviceEndpointListener deviceEndpointListener) {
        canceled = false;
        logger.info("starting discovery");
        if (discovery == null) {
            lock.acquire();
            discovery = new NsdDiscovery(manager, deviceEndpointListener, "syncloud");
            discovery.start();
            try {
                logger.info("waiting for " + timeoutSeconds + " seconds");
                int count = 0;
                while(count < timeoutSeconds && !canceled) {
                    Thread.sleep(1000);
                    count++;
                }
            } catch (InterruptedException e) {
                logger.error("sleep interrupted", e);
            }
            stop();
        }
    }

    public void cancel() {
        canceled = true;
    }

    private void stop() {
        logger.info("stopping discovery");
        if (discovery != null) {
            try {
                discovery.stop();
                discovery = null;
            } catch (Exception e) {
                logger.error("failed to stop discovery", e);
            }
            lock.release();
        }
    }

}

File: syncloud/src/main/java/org/syncloud/android/discovery/DiscoveryManager.kt
package org.syncloud.android.discovery

import android.net.nsd.NsdManager
import android.net.wifi.WifiManager
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.syncloud.android.Logger
import org.syncloud.android.discovery.nsd.NsdDiscovery
import java.util.Collections

const val SERVICE_NAME = "syncloud"

class DiscoveryManager(wifi: WifiManager, private val manager: NsdManager) {
    private val lock: MulticastLock = MulticastLock(wifi)
    private var discovery: Discovery? = null
    private var canceled = false

    suspend fun run(timeoutSeconds: Int, added: suspend (device: String) -> Unit) {
        canceled = false
        logger.info("starting discovery")
        if (discovery != null) return

        val reported = Collections.synchronizedSet(mutableSetOf<String>())
        val report: suspend (String) -> Unit = { device ->
            if (reported.add(device)) added(device)
        }

        lock.acquire()
        discovery = NsdDiscovery(manager, report, SERVICE_NAME)
        (discovery as NsdDiscovery).start()
        coroutineScope {
            launch {
                UnicastDiscovery(SERVICE_NAME).query(timeoutSeconds, { canceled }, report)
            }
            logger.info("waiting for $timeoutSeconds seconds")
            var count = 0
            while (count < timeoutSeconds && !canceled) {
                delay(1000)
                count++
            }
        }
        stop()
    }

    fun cancel() {
        canceled = true
    }

    private fun stop() {
        logger.info("stopping discovery")
        if (discovery != null) {
            try {
                discovery?.stop()
                discovery = null
            } catch (e: Exception) {
                logger.error("failed to stop discovery", e)
            }
            lock.release()
        }
    }

    companion object {
        private val logger = Logger.getLogger(DiscoveryManager::class.java.name)
    }
}
